src/combinatorics/hales_jewett.lean

```diff
@@ -302,7 +308,78 @@ begin
     intros i hi,
     rw [hs, finset.sep_def, finset.compl_filter, finset.mem_filter] at hi,
     obtain ⟨y, hy⟩ := option.ne_none_iff_exists.mp hi.right,
-    simp_rw [line.apply, ←hy, option.map_some', option.get_or_else_some], },
+    rw [l.apply_some hy.symm, ←hy, option.map_some', option.get_or_else_some] }
+end
+
+/-- `subspace η α ι` is the type of `η`-dimensional *combinatorial subspaces* of `ι → α`. We do not
+allow degenerate subspaces. -/
+structure subspace (η α ι : Type*) :=
+(idx_fun : ι → α ⊕ η)
+(proper : ∀ e, ∃ i, idx_fun i = sum.inr e)
+
+namespace subspace
+
+/-- The combinatorial subspace corresponding to the identity embedding `(ι → α) → (ι → α)`. -/
+instance {α ι} : inhabited (subspace ι α ι) := ⟨⟨sum.inr, λ i, ⟨i, rfl⟩⟩⟩
+
+instance (η α ι) : has_coe_to_fun (subspace η α ι) (λ _, (η → α) → ι → α) :=
+⟨λ l x i, (l.idx_fun i).elim id x⟩
+
+lemma apply_def {η α ι} (l : subspace η α ι) (x : η → α) (i : ι) :
+  l x i = (l.idx_fun i).elim id x := rfl
+
+lemma apply_inl {η α ι} {l : subspace η α ι} {x : η → α} {i : ι} {y : α}
+  (h : l.idx_fun i = sum.inl y) : l x i = y :=
+by rw [l.apply_def, h, sum.elim_inl, id.def]
+
+lemma apply_inr {η α ι} {l : subspace η α ι} {x : η → α} {i : ι} {j : η}
+  (h : l.idx_fun i = sum.inr j) : l x i = x j :=
+by rw [l.apply_def, h, sum.elim_inr]
+
+/-- Given a coloring `C` of `ι → α` and a combinatorial subspace `l` of `ι → α`, `l.is_mono C` means
+that `l` is monochromatic with regard to `C`. -/
+def is_mono {η α ι κ} (C : (ι → α) → κ) (l : subspace η α ι) : Prop :=
+∃ c, ∀ x, C (l x) = c
+
+/-- The **extended Hales-Jewett theorem**. We deduce it from the ordinary Hales-Jewett theorem. -/
```

Comment:
I think the docstring should describe what the theorem says, not so much how it is proven. Can you please give a 1 or 2 line description of what the theorem says?